### DISCUSSION TITLE
Concurrence Management with _etag

### DISCUSSION THREAD

**Discussion by TarekKassem** [Gremlinq/ExRam.Gremlinq#600] [Q&A] [answered]
Hi @danielcweber , 
I am implementeing Exram.Gremlinq with cosmos graph db, and I have an issue concerning the concurrence when doing updates, I saw that the azure cosmos client has a feature "ifMatchEtag"  that is based on the _etag property generated by the cosmos database. I wanted to know if there is something similar in Exram.Gremlinq that allows to manage this issue ?
Thanks for the great wok on achieving this library by the way !

**Comment by danielcweber** [comment, marked as answer]
That feature is only available on CosmosDbs document (SQL) API. There's nothing similar on the Gremlin API and thus, also not in ExRam.Gremlinq.

**Reply by TarekKassem** [reply]
Thanks for your answer

**Reply by TarekKassem** [reply]
Did you have issues with concurrency when doing updates and deletes when you implelented Exram.Gremlinq ? And if yes how did you manage these cases ? I am thinking of adding a custom property on my vertices which contains a timestamp as a value and to do a check on this vaue when doing updates and deletes to be sure that I have the latest version of my vertex at the moment of applying the update/delete. Any suggestions are welcome!

**Reply by danielcweber** [reply]
You'd need the etag to be sure you're updating exactly the document you just read, and there's only the document API for this. No guarantees on the Gremlin API. Your approach might work 99% of the time, and the remaining 1% will give you a hard time, depending on how critical these kinds of changes are to you.

**Reply by danielcweber** [reply]
Check

https://docs.microsoft.com/en-us/azure/cosmos-db/sql/database-transactions-optimistic-concurrency and
https://docs.microsoft.com/en-us/azure/cosmos-db/sql/how-to-manage-consistency

**Reply by TarekKassem** [reply]
ok I see, thank you Daniel
